lib/referencer.js

```diff
@@ -453,6 +459,17 @@ class Referencer extends esrecurse.Visitor {
         this.visitProperty(node);
     }
 
+    PropertyDefinition(node) {
+        if (node.computed) {
+            this.visit(node.key);
+        }
+        if (node.value) {
+            this.scopeManager.__nestFunctionScope(node, /* strict= */ true);
+            this.visit(node.value);
+            this.close(node);
+        }
+    }
```

Comment:
This correctly handles scope-reference relations, but it looks unusual that a node that creates a scope can contain parts that belong to its upper scope. For example, it's unexpected for [context.getScope()](https://github.com/eslint/eslint/blob/aa87329d919f569404ca573b439934552006572f/lib/linter/linter.js#L710) and it will return a wrong scope when called from within a computed key:

```js
class A {
    [x] = x;
}
```

```js
"Identifier[name='x']"() {
    context.getScope().type // "function" "function";
}
```

Reply:
@mdjermanovic what would you suggest instead? Let’s make sure to keep comments actionable so we can move things along quickly.

Reply:
Indeed. The implicit function is the class field initializer, but not whole the class field. Probably the function scope is only the initializer node.

Reply:
It's difficult to predict the impact on the existing logic in rules.

`scope.block = PropertyDefinition` is surprising because of the tree hierarchy.
`scope.block = PropertyDefinition.value` is surprising because any expression can create a scope.

If the AST spec had a function node between property and its value, that would be probably the least surprising.

I think `scope.block = PropertyDefinition.value` is logically correct choice.

Reply:
> I think `scope.block = PropertyDefinition.value` is logically correct choice.

Also, if we decide on this approach, would it be useful to have a different `scope.type` (e.g., "class-field-initializer") to make a distinction between the initializer scope and the scope that the value itself creates in some cases?

For example, if we use `"function"`, then the ArrowFunctionExpression in the following example has two `"function"` scopes:

```js
class A {
    x = () => y;
}
```

Reply:
> It's difficult to predict the impact on the existing logic in rules.

I agree.

I have introduced the `class-field-initializer` scope.
Though we cannot put variable declarations into the scope, the scope's `variableScope` is the scope itself, as the same as function scopes. Because some rules use `variableScope`'s equality to detect being in the same function, it will be better.

Because `context.getScope(node)` for function nodes returns the innermost scope for the function, `context.getScope(node)` for function nodes never returns the new `class-field-initializer` scope.

Reply:
> Though we cannot put variable declarations into the scope, the scope's `variableScope` is the scope itself, as the same as function scopes. Because some rules use `variableScope`'s equality to detect being in the same function, it will be better.

I agree, the [documentation](https://github.com/eslint/eslint/blob/master/docs/developer-guide/scope-manager-interface.md#variablescope) says it's about `var` variables, but the practical use is to find the nearest enclosing function scope. This would match the expectations of rules in https://github.com/eslint/eslint/pull/14591#issuecomment-866168700.

Reply:
Except for static fields, which seem to be a special case as they are initialized right after the class definition. From the perspective of `eslint-scope` I think it makes sense to treat static fields the same as instance fields, but rules will have to handle them specifically, if needed.